test/v2/services/RetrieveBalanceAndTransactionsServiceSpec.scala

```diff
@@ -0,0 +1,108 @@
+import v2.fixtures.retrieveBalanceAndTransactions.DocumentDetailsFixture.documentDetails
+import v2.fixtures.retrieveBalanceAndTransactions.FinanceDetailsFixture.financeDetailsFullObject
+import v2.mocks.connectors.MockRetrieveBalanceAndTransactionsConnector
+import v2.models.request.retrieveBalanceAndTransactions.RetrieveBalanceAndTransactionsRequest
+import v2.models.response.retrieveBalanceAndTransactions._
+import v2.services.RetrieveBalanceAndTransactionsService.downstreamErrorMap
+
+import scala.concurrent.Future
+
+class RetrieveBalanceAndTransactionsServiceSpec extends ServiceSpec {
+
+  private val nino = "AA123456A"
+  private val docNumber = "anId"
+  private val dateFrom = "2018-08-13"
+  private val dateTo = "2018-08-14"
+  private val onlyOpenItems = false
+  private val includeLocks = false
+  private val calculateAccruedInterest = false
+  private val removePOA = false
+  private val customerPaymentInformation = false
+  private val includeStatistical = false
+
+  private val requestData: RetrieveBalanceAndTransactionsRequest =
+    RetrieveBalanceAndTransactionsRequest(
+      Nino(nino),
+      Some(docNumber),
+      Some(dateFrom),
+      Some(dateTo),
+      onlyOpenItems,
+      includeLocks,
+      calculateAccruedInterest,
+      removePOA,
+      customerPaymentInformation,
+      includeStatistical)
+
+  val retrieveBalanceAndTransactionsResponse: RetrieveBalanceAndTransactionsResponse =
+    RetrieveBalanceAndTransactionsResponse(
+      balanceDetails = balanceDetailsObject,
+      codingDetails = Some(Seq(codingDetailsObject)),
+      documentDetails = Some(Seq(documentDetails)),
+      financeDetails = Some(Seq(financeDetailsFullObject))
+    )
+
+  "RetrieveBalanceAndTransactionsService" should {
+    "service call successful" when {
+      "return mapped result" in new Test {
+        MockRetrieveBalanceAndTransactionsConnector
+          .retrieveBalanceAndTransactions(requestData)
+          .returns(Future.successful(Right(ResponseWrapper(correlationId, retrieveBalanceAndTransactionsResponse))))
+
+        await(service.retrieveBalanceAndTransactions(requestData)) shouldBe Right(ResponseWrapper(correlationId, retrieveBalanceAndTransactionsResponse))
+      }
+    }
+
+    "map errors according to spec" when {
+      def serviceError(downstreamErrorCode: String, error: MtdError): Unit =
+        s"a $downstreamErrorCode error is returned from the service" in new Test {
+
+          MockRetrieveBalanceAndTransactionsConnector
+            .retrieveBalanceAndTransactions(requestData)
+            .returns(Future.successful(Left(ResponseWrapper(correlationId, DownstreamErrors.single(DownstreamErrorCode(downstreamErrorCode))))))
+
+          private val result = await(service.retrieveBalanceAndTransactions(requestData))
+          result shouldBe Left(ErrorWrapper(correlationId, error))
+        }
+
+      downstreamErrorMap.foreach(args => (serviceError _).tupled(args))
```

Comment:
I feel like it's safer to duplicate the error map in the test, rather than import it from the service, as if there is an error in the mappings, the test wouldn't catch it. E.g. we could delete all the error codes and have an empty map, and this test would continue to pass.